File: 동적계획법/계단오르기.py
import sys
# sys.stdin = open("input.txt", 'r')

def dfs(n):
    if memo[n] != 0:
        return memo[n]

    if n == 1 or n == 2:
        return n
    else:
        memo[n] = dfs(n - 1) + dfs(n - 2)
        return memo[n]

if __name__ == "__main__":
    n = int(input())
    memo = [0] * (n + 1)
    dfs(n)
    print(memo[n])

File: 동적계획법/계단오르기.cpp
/*
#include <bits/stdc++.h>
using namespace std;

int main(void){
	ios_base::sync_with_stdio(false);	
	cin.tie(NULL);
	freopen("input.txt", "r", stdin);
	
	int n;
	int i;
	cin >> n;
	vector<int> length(n + 1);
	
	length[1] = 1;
	length[2] = 2;
	for(i = 3; i <= n; i++){
		length[i] = length[i - 1] + length[i - 2];	
	}
	cout << length[n];
	
	return 0;
}
*/
#include <bits/stdc++.h>
using namespace std;

vector<int> memo(46);

int dfs(int len){
	if(memo[len] != 0) return memo[len];
	
	if(len == 1 || len == 2) return len;
	else{
		return memo[len] = dfs(len - 1) + dfs(len - 2);
	}
}

int main(void){
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	// freopen("input.txt", "r", stdin);
	
	int n;
	scanf("%d", &n);
	cout << dfs(n);
		
	return 0;
}
